base/strings/string.jl

```diff
@@ -190,14 +191,71 @@ end
 end
 
 ## checking UTF-8 & ACSII validity ##
+# This table is used by the shift base DFA validation for UTF-8
+const _UTF8_DFA_TABLE = [
+    fill(UInt64(109802048057794944), 128);
+    fill(UInt64(113232530780455302), 16);
+    fill(UInt64(109855655693648262), 16);
+    fill(UInt64(109855649351860614), 32);
+    fill(UInt64(109802048057794950), 2);
+    fill(UInt64(109802048057794956), 30);
+    fill(UInt64(109802048057794968), 1);
+    fill(UInt64(109802048057794962), 12);
+    fill(UInt64(109802048057794974), 1);
+    fill(UInt64(109802048057794962), 2);
+    fill(UInt64(109802048057794980), 1);
+    fill(UInt64(109802048057794986), 3);
+    fill(UInt64(109802048057794992), 1);
+    fill(UInt64(109802048057794950), 11)
+    ]::Vector{UInt64}
+
+const _UTF8_DFA_ACCEPT = UInt64(0) #This state represents the start and end of any valid string
+const _UTF8_DFA_INVALID = UInt64(6) # If the state machine is ever in this state just stop
+
+# This function is designed so that you could use it on strings with discontinous memmory layouts
+#   by only feeding it contiguous block and keeping track of the state inbetween. 
+# Furthermore you could check in returned value is _UTF8_DFA_INVALID and stop as invalid if it was.
+# For a contiguous bytestream other states are valid other than _UTF8_DFA_ACCEPT aslong as you aren't
+#  at the begining or end
+function _isvalid_utf8_dfa(bytes::Vector{UInt8},state::UInt64 = _UTF8_DFA_ACCEPT)
+    f(byte) = @inbounds _UTF8_DFA_TABLE[byte+1]
+    op(s, byte_dfa) = byte_dfa >> (s & UInt64(63))
+    final_state = mapfoldl(f, op, bytes, init = state)
+    return (final_state & UInt64(63)) 
+end
+
+# This is a shift based utf-8 DFA that works on string that are a contiguous block
+function _isvalid_utf8(bytes::Vector{UInt8})
+    final_state = _isvalid_utf8_dfa(bytes, _UTF8_DFA_ACCEPT)
+    return (final_state & UInt64(63)) == _UTF8_DFA_ACCEPT
+end
 
-byte_string_classify(s::Union{String,Vector{UInt8},FastContiguousSubArray{UInt8,1,Vector{UInt8}}}) =
-    ccall(:u8_isvalid, Int32, (Ptr{UInt8}, Int), s, sizeof(s))
+_isvalid_utf8(s::Union{String,FastContiguousSubArray{UInt8,1,Vector{UInt8}}}) = _isvalid_utf8(unsafe_wrap(Vector{UInt8}, s))
+
+# Classifcations of string
     # 0: neither valid ASCII nor UTF-8
     # 1: valid ASCII
     # 2: valid UTF-8
+function byte_string_classify(s::Union{String,FastContiguousSubArray{UInt8,1,Vector{UInt8}}})
+    bytes = unsafe_wrap(Vector{UInt8}, s)
+    byte_string_classify(bytes, kwargs...)
+end
+
+function byte_string_classify(bytes::Vector{UInt8})
+    all(c -> iszero(c & 0x80), bytes) && return 1
+    valid = _isvalid_utf8(bytes)
+    return ifelse(valid, 2, 0)
+end
+
+function isvalid(::Type{String}, s::Union{FastContiguousSubArray{UInt8,1,Vector{UInt8}},String}) 
+    bytes = unsafe_wrap(Vector{UInt8}, s)
+    isvalid(String,bytes)
+end
+function isvalid(::Type{String}, bytes::Vector{UInt8}) 
+    valid = _isvalid_utf8(bytes)
+    return ifelse(valid, true, false)
```

Comment:
isn't this just `isvalid(::Type{String}, bytes::Vector{UInt8}) = _isvalid_utf8(bytes)`

Reply:
They are, but this one of those areas that is very sensitive to what optimations it gets.  On my machine that is the difference between 2.2 GB/s and 1.5 GB/s.

Reply:
then your benchmark is wrong. Is the `@code_native` different?

Reply:
I checked it and you were right the native code was the same after I added an `@inline`